enhancements/art/flexibly-enforcing-product-build-constraints-in-ci.md

```diff
@@ -0,0 +1,266 @@
+    3. The alignment PR is updated automatically to reflect the new desired state. If that desired state 
+       is identical to Day 0, the alignment PR is a no-op and can be closed.
+6. **Day 45**: Feature freeze is approaching. ART queries for open alignment PRs and finds 2 in the search. 
+   Urgent BZs are opened against both components to communicate the need for either alignment or production metadata changes.
+7. **Day 48**: Feature freeze. ART changes production metadata to reflect the beginning of a new stage of the software 
+   lifecycle. One urgent BZ, despite escalation, remains unanswered. It is changed to a blocker bug. To ensure consistent 
+   and predictable builds leading up to and after GA, the production metadata is now authoritative and the upstream 
+   discrepancy is ignored (the PR and BZ remain open).
+8. **Day N**: If, at any point after feature freeze, the ART build pipeline detects that an upstream repository is not using 
+   the builder image defined in ART metadata, the release will be marked as inconsistent (visible in the release controller 
+   and programmatically accessible by checking an annotation). If the ART pipeline attempts to promote an inconsistent release, 
+   it will abort unless it is manually overridden. The manual override would only be undertaken if it was determined, 
+   with architectural approval, that the inconsistency was introduced erroneously and that the promotion is a critical 
+   delivery (e.g. a CVE fix).
+   
+### Implementation Details/Notes/Constraints
+
+If an upstream component owner does not just accept PRs (e.g. selects an arbitrary go version or selects 
+an out of support ART equivalent from several releases ago), ART will not be able to reverse lookup the product 
+equivalent image. The component owner may believe they are building using a particular version of golang in 
+production, while they are not. It should be noted that, during this time, an alignment PR will remain open.
+
+The proposal also implies that CI will make use of the FROM lines in the repository Dockerfiles. CI can 
+and will apply a mapping to these, but will derive the Golang version (or other versions of that matter) 
+from the repository files. The case of explicit image overrides in the openshift/release repository will become the exception, 
+i.e. a PR against the upstream repository will be enough to switch the version of the base or build image.
+
+When the ART pipeline detects that a given build has inconsistent upstream and downstream configurations, 
+it will still feed the build into the OpenShift release controller. However, the imagestream tag within 
+the 4.x-art-latest imagestream (which the release controller consumes) will be annotated with the reason 
+ART believes the subsequent release cannot be used for a production. The release controller will propagate 
+each inconsistency annotation within the ART imagestream into the subsequent release payload's imagestream 
```

Comment:
Will the presence of the `release.openshift.io/inconsistency` annotation in a nightly block nightly acceptance, just like sufficient promotion-gate job failures block nightly acceptance today?

Reply:
It will not block acceptance, but it will block promotion. QE will be told that, when selecting a nightly for release, they should not consider an inconsistent release. The ART pipeline will also refuse to promote an inconsistent release unless it is manually overridden.

Reply:
So inconsistent releases still get accepted and some subset of them are mirrored to Quay as nightlies.  But by the time we start cutting Feature Candidates, we have resolved everything one way or the other.  Sounds good to me.  Is this too tangential to include in the enhancement text?

Reply:
Yes, I'll add it. Thanks!

Reply:
See `An "inconsistent" release will still be accepted by ...`